### PR TITLE
checkpoint `reshape_text` - partial fix

### PR DESCRIPTION
**Pull request by t-bltg**
@FedeClaudi, I think `reshape_text` is not very robust. I give up now since I spent too much time on this.

- indexing into generic string with possible non-unit byte characters should be exerted with caution and using `nextindex(...)` and friends (see https://docs.julialang.org/en/v1/manual/strings/#Unicode-and-UTF-8) - it's always a mess working with those strings;
- `01_test_text_utils.jl` passes now with some limitations, and added test cases;
- ~~all others test need manual fixes when comparing output strings~~ ==> all tests green now.

The
```julia
    for i in 5:10
        width = 2^i
        for offset in -10:10
            txt = '.'^(2width + offset)
            rt = reshape_text(txt, width)
            len = length.(split(rt, '\n'))
            # @show length(txt) width rt len
            @test all(len[1:(end - 1)] .== width)
            @test len[end] == (offset > 0 ? offset : width + offset)
        end
    end
```

test covers fixing the issue mentioned at https://github.com/FedeClaudi/Term.jl/pull/102#issuecomment-1153611730.

### CODE REVIEW

File: test/01_test_text_utils.jl [hunk 1]
```diff
@@ -172,47 +175,60 @@ id est laborum."""
...
 
-    for (s1, s2) in strings
-        reshaped = reshape_text(s1, 33)
-        @test reshaped == s2
-    end
-
-    for width in (40, 60, 99)
-        rh = reshape_text(str, width)
-        @test !any(textlen.(split(rh, "\n")) .> width)
+    width = 33
```
Comment (FedeClaudi) [inline test/01_test_text_utils.jl:212]: Can you walk me through what's happening in these lines?

**Comment by FedeClaudi** [comment]
> I give up now since I spent too much time on this.

Of course, I appreciate your help.

> I think reshape_text is not very robust

Reshaping text has been a major headache.. I've changed the way it works multiple times but yeah it's still not great. It needs to:
1. work decently efficiently
2. insert new lines at spaces without breaking words (unless the word is too long)
3. ignore ANSI and markup style info
4. ensure that style is carried over across multiple lines without bleeding into other content (e.g. when horizontally stacking two pieces of text..)

and of course the way strings are handled in Julia is a pain... 
> ...non-unit byte characters...

which includes e.g. Panel's borders, so this happens all the time. But I've found that using `nextindex` on strings that don't have such characters gives some unexpected results... I don't understand why Julia handles strings the way it does instead of allowing indexing to access individual characters. 


> all tests green now

You're a hero! I'm very impressed with all the changes you've been making!

File: test/01_test_text_utils.jl (hunk 1 above)
Comment (t-bltg) [reply test/01_test_text_utils.jl:212]: Ha sorry, they were moved above the `enumerate(strings)` test. I'm going to restore the testing order ...

**Comment by t-bltg** [comment]
> Reshaping text has been a major headache..

Gave me some too :laughing: .

> I don't understand why Julia handles strings the way it does instead of allowing indexing to access individual characters.

Very special yes :hot_face: .

> But I've found that using nextindex on strings that don't have such characters gives some unexpected results

Let's reserve that for future improvements, this limitation causes strings such as `ต้าอ่วยวาทกรรมอาว์เซี้ยว กระดี๊กระด๊า ช็อปซาดิสต์โมจิดีพาร์ตเมนต์ อินดอร์วิว สี่แยกมาร์กจ๊อกกี้ โซนี่บัตเตอร์ฮันนีมูน ยาวีแพลนหงวนสคริปต์ แจ็กพ็อตต่อรองโทรโข่งยากูซ่ารุมบ้า บอมบ์เบอร์รีวีเจดีพาร์ทเมนท์ บอยคอตต์เฟอร์รี่บึมมาราธอน ` to **not** respect the width parameter when split using `reshape_text`.

**Comment by t-bltg** [comment]
> that was working for me before

Was it ? The split lines in https://github.com/FedeClaudi/Term.jl/blob/a2a46591f79f87b6269bfc26aebc5098411aefd4/test/01_test_text_utils.jl#L159 have `length = [38, 39, 41, ...]`, thus not respecting `width = 33` given as arg.

**Comment by FedeClaudi** [comment]
mmm the tests were passing, we maybe the tests weren't doing what I wanted them to do 😄 
either way I agree that's not a high priority for now

**Comment by FedeClaudi** [comment]
Okay, looks good. Happy to merge when you're ready!

**Comment by t-bltg** [comment]
> Happy to merge when you're ready!

Just bumping the coverage across the repo, and we should be good to go in a few minutes ...

**Comment by t-bltg** [comment]
Good to merge when CI is green :tada: !

**Pull request merged**